File: src/python/dynamic_programming/2293/main.py
# Date : 2017.11.11
# Author : seungyeon choi
# 2293 (동전1)

N, K = map(int, input().split())
coin = []
DP = [0 for _ in range(10001)]
DP[0] = 1
for _ in range(N):
    coin.append(int(input()))
for i in range(N):
    for j in range(coin[i], K+1, 1):
        DP[j] += DP[j-coin[i]]
print (DP[K])

File: src/c++/dynamic_programming/2293/main.cc
/* Date : 2019.5.18
 * Author : seungyeon choi
 * 2293 (동전1)
 */

#include <cstdio>

using namespace std;

const int MAX_COIN = 101;
const int MAX_COUNT = 10001;

int N, K, ans;
int COIN[MAX_COIN];
int DP[MAX_COUNT];

int main() {
  scanf("%d %d", &N, &K);

  for (int i = 0; i < N; i++)
    scanf("%d", &COIN[i + 1]);

  DP[0] = 1;

  for (int i = 1; i <= N; i++) {
    for (int j = COIN[i]; j <= K; j++) {
      if (j >= COIN[i])
        DP[j] += DP[j - COIN[i]];
    }
  }

  printf("%d\n", DP[K]);

  return 0;
}
